### PR TITLE
Clarify `fast_dev_run` docs

### LINKED ISSUE #12738: [RFC] Is `fast_dev_run` still needed?
### Motivation

This was feedback from a user:

> Hi, we found an unexpected error when training a model with fast_dev_run=True. The training finishes with single train and val steps, but writing a further outputs to the checkpoint directory fails. 

> How is fast_dev_run mode different from just setting to 1 epoch and 1 batch per epoch? My understanding is that fast_dev_run is almost equivalent to `limit_train_batches=1` and `limit_val_batches=1` (or something like that), but when I set these params instead of fast_dev_run there were no issues at all.

The issue was that within the checkpoint callback, we do not create a checkpoint directory if `fast_dev_run` is used. It is not obvious at all to users what this flag does based on the documentation. I had to grep through the Lightning codebase to see where all `fast_dev_run` takes effect. The ambiguity makes it hard to reason about. The framework is also inconsistent with this flag: should the model summary be printed? Should early stopping take effect? What about loggers or profilers? Why some but not others?

We already have Trainer flags for `max_steps`, `limit_train_batches`, `enable_checkpointing`, etc. Do we still need `fast_dev_run` on top of this?

### Pitch

Deprecate `fast_dev_run` in favor of the specific flags already offered on the Trainer?

### Additional context

______________________________________________________________________

#### If you enjoy Lightning, check out our other projects! ⚡

- [**Metrics**](https://github.com/PyTorchLightning/metrics): Machine learning metrics for distributed, scalable PyTorch applications.

- [**Lite**](https://pytorch-lightning.readthedocs.io/en/latest/starter/lightning_lite.html): enables pure PyTorch users to scale their existing code on any kind of device while retaining full control over their own loops and optimization logic.

- [**Flash**](https://github.com/PyTorchLightning/lightning-flash): The fastest way to get a Lightning baseline! A collection of tasks for fast prototyping, baselining, fine-tuning, and solving problems with deep learning.

- [**Bolts**](https://github.com/PyTorchLightning/lightning-bolts): Pretrained SOTA Deep Learning models, callbacks, and more for research and production with PyTorch Lightning and PyTorch.

- [**Lightning Transformers**](https://github.com/PyTorchLightning/lightning-transformers): Flexible interface for high-performance research using SOTA Transformers leveraging Pytorch Lightning, Transformers, and Hydra.

cc @borda @justusschock @kaushikb11 @awaelchli @rohitgr7

### PR DESCRIPTION
**Pull request by carmocca**
## What does this PR do?

Fixes #12738

### Does your PR introduce any breaking changes? If yes, please list them.

None

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [n/a] Did you write any **new necessary tests**? (not for typos and docs)
- [n/a] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [n/a] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

cc @borda @rohitgr7 @justusschock @kaushikb11 @awaelchli

### CODE REVIEW

File: docs/source/common/trainer.rst [hunk 1]
```diff
@@ -669,43 +669,38 @@ fast_dev_run
 
 |
 
-Runs n if set to ``n`` (int) else 1 if set to ``True`` batch(es) of train, val and test
-to find any bugs (ie: a sort of unit test).
```
Comment (carmocca) [inline docs/source/common/trainer.rst:673]: The unit test comment is very misleading considering it doesn't run any special assertions

File: docs/source/common/trainer.rst [hunk 2]
```diff
@@ -669,43 +669,38 @@ fast_dev_run
 
 |
 
-Runs n if set to ``n`` (int) else 1 if set to ``True`` batch(es) of train, val and test
```
Comment (carmocca) [inline docs/source/common/trainer.rst:672]: The wording here made it sound like `trainer.fit()` would still run 1 test batch. Not the case

**Review by rohitgr7 — APPROVED** [review]

File: docs/source/common/trainer.rst [hunk 3]
```diff
@@ -669,43 +669,38 @@ fast_dev_run
...
-.. note::
-
-    This argument is a bit different from ``limit_train/val/test_batches``. Setting this argument will
-    disable tuner, checkpoint callbacks, early stopping callbacks, loggers and logger callbacks like
-    ``LearningRateLogger`` and runs for only 1 epoch. This must be used only for debugging purposes.
-    ``limit_train/val/test_batches`` only limits the number of batches and won't disable anything.
+    trainer.predict(...)
+
+This argument is a different from ``limit_{train,val,test,predict}_batches`` because side effects are avoided to reduce the
+impact to subsequent runs. Here are the key differences:
```
Comment (carmocca) [inline docs/source/common/trainer.rst:690]:
```suggestion
This argument is different from ``limit_{train,val,test,predict}_batches`` because side effects are avoided to reduce the
impact to subsequent runs. These are the changes enabled:
```

**Review by ananthsub — APPROVED** [review]
Thanks @carmocca !

**Review by Borda — APPROVED** [review]

**Review by justusschock — APPROVED** [review]

**Review by akihironitta — APPROVED** [review]

**Review by awaelchli — APPROVED** [review]

**Review by kaushikb11 — APPROVED** [review]
